riscv-runtime/src/hash.rs

```diff
@@ -40,79 +40,100 @@ pub fn poseidon2_gl(data: &[Goldilocks; 8]) -> [Goldilocks; 8] {
-    buffer: [u8; 136],
-    buffer_pos: usize,
-    rate: usize,
+    state: [u32; 50],
+    next_word: usize,
+    input_buffer: u32,
+    next_byte: usize,
 }
 
 impl Keccak {
-    pub fn v256() -> Self {
-        //RATE = 136
+    const RATE: usize = 34; // Rate in u32 words
 
+    pub fn v256() -> Self {
         Self {
-            state: [0u64; 25],
-            buffer: [0u8; 136],
-            buffer_pos: 0,
-            rate: 136,
+            state: [0u32; 50],
+            next_word: 0,
+            input_buffer: 0,
+            next_byte: 0,
         }
     }
 
-    pub fn update(&mut self, data: &[u8]) {
-        let mut input = data;
-        while !input.is_empty() {
-            let space = self.rate - self.buffer_pos;
-            let to_absorb = space.min(input.len());
-
-            self.buffer[self.buffer_pos..self.buffer_pos + to_absorb]
-                .copy_from_slice(&input[..to_absorb]);
-            self.buffer_pos += to_absorb;
-            input = &input[to_absorb..];
-
-            if self.buffer_pos == self.rate {
-                self.absorb_block();
-                self.buffer_pos = 0;
-            }
+    fn xor_word_to_state(&mut self, word: u32) {
+        let word_pair = self.next_word & !1;
+        if (self.next_word & 1) == 0 {
+            self.state[word_pair + 1] ^= word;
+        } else {
+            self.state[word_pair] ^= word;
         }
-    }
+        self.next_word += 1;
 
-    fn absorb_block(&mut self) {
-        let rate_words = self.rate / 8;
-        for i in 0..rate_words {
-            let chunk = &self.buffer[i * 8..(i + 1) * 8];
-            let val = u64::from_le_bytes(chunk.try_into().unwrap());
-            // Note: lo/hi positions are swapped to match the Keccak machine memory layout
-            let swapped = ((val << 32) & 0xFFFFFFFF00000000) | ((val >> 32) & 0x00000000FFFFFFFF);
-            self.state[i] ^= swapped;
+        if self.next_word == Self::RATE {
+            let mut state_out = [0u32; 50];
+            keccakf(&self.state, &mut state_out);
+            self.state = state_out;
+            self.next_word = 0;
         }
+    }
+
+    pub fn update(&mut self, data: &[u8]) {
+        unsafe {
+            let (prefix, words, suffix) = data.align_to::<u32>();
+
+            for &byte in prefix {
+                self.input_buffer |= (byte as u32) << (8 * self.next_byte);
+                self.next_byte += 1;
+                if self.next_byte == 4 {
+                    self.xor_word_to_state(self.input_buffer);
+                    self.input_buffer = 0;
+                    self.next_byte = 0;
+                }
+            }
+
+            for &word in words {
+                if self.next_byte == 0 {
+                    self.xor_word_to_state(word);
+                } else {
+                    self.xor_word_to_state(self.input_buffer | (word << (8 * self.next_byte)));
+                    self.input_buffer = word >> (32 - 8 * self.next_byte);
+                }
```

Comment:
Ah, pure poetry!

It is nice that you treated the special case `self.next_byte == 0` differently. I probably wouldn't, but I like it.

It would execute faster if the `if` was outside the `for`, but I am pretty confident LLVM can optimize it, and this way it is easier to read.